crates/publish/src/oci/client.rs

```diff
@@ -0,0 +1,288 @@
+        let reference: Reference = reference.parse().context("cannot parse reference")?;
+        let auth = Self::auth(&reference)?;
+
+        // Pull the manifest from the registry.
+        let (manifest, digest) = self.oci.pull_image_manifest(&reference, &auth).await?;
+
+        let manifest_json = serde_json::to_string(&manifest)?;
+        tracing::debug!("Pulled manifest: {}", manifest_json);
+
+        // Write the manifest in `<cache_root>/registry/oci/manifests/repository:<tag_or_latest>/manifest.json`
+        let m = self
+            .cache
+            .manifest_for_reference(&reference.to_string())
+            .await?;
+        fs::write(&m, &manifest_json).await?;
+
+        let mut cfg_bytes = Vec::new();
+        self.oci
+            .pull_blob(&reference, &manifest.config.digest, &mut cfg_bytes)
+            .await?;
+        let cfg = std::str::from_utf8(&cfg_bytes)?;
+        tracing::debug!("Pulled config: {}", cfg);
+
+        // Write the config object in `<cache_root>/registry/oci/manifests/repository:<tag_or_latest>/config.json`
+        let c = self
+            .cache
+            .config_for_reference(&reference.to_string())
+            .await?;
+        fs::write(&c, &cfg).await?;
+
+        // If a layer is a Wasm module, write it in the Wasm directory.
+        // Otherwise, write it in the data directory.
+        for layer in manifest.layers {
+            // Skip pulling if the digest already exists in the wasm or data directories.
+            if std::fs::metadata(&self.cache.wasm_dir().join(&layer.digest)).is_ok()
+                || std::fs::metadata(&self.cache.data_dir().join(&layer.digest)).is_ok()
+            {
+                tracing::debug!("Layer {} already exists in cache", &layer.digest);
+                continue;
+            }
+            tracing::debug!("Pulling layer {}", &layer.digest);
+            let mut bytes = Vec::new();
+            self.oci
+                .pull_blob(&reference, &layer.digest, &mut bytes)
+                .await?;
+
+            match layer.media_type.as_str() {
+                WASM_LAYER_MEDIA_TYPE => self.cache.write_wasm(&bytes, &layer.digest).await?,
+                _ => self.cache.write_data(&bytes, &layer.digest).await?,
+            }
+        }
+
+        tracing::info!("Pulled {}@{}", reference, digest);
+
+        Ok(())
+    }
+
+    pub async fn descriptor(&mut self, reference: &str) -> Result<SpinOciDescriptor> {
+        let reference: Reference = reference.parse().context("cannot parse reference")?;
+
+        let manifest_path = self
+            .cache
+            .manifest_for_reference(&reference.to_string())
+            .await?;
+        let config_path = self
+            .cache
+            .config_for_reference(&reference.to_string())
+            .await?;
+        if !manifest_path.exists() || !config_path.exists() {
+            self.pull(&reference.to_string()).await?;
```

Comment:
The function name here made me think that it computed a descriptor - given that pulling is a heavyweight operation, could the name imply that it will pull if necessary?

(I'm not yet sure what this function is used for, so I don't have any suggestions yet, sorry.)

Reply:
Easy fix for this one — the function was not called anymore, so removed it altogether.